README.md
A unifified manipulation of time and date representations.

Uniform means:
- same functions with identical semantics independent of representation
- all functions are total (or become so using Maybe or Either)

Conversions from `EpochTime` to `UTCTime` and then to `YMD` and similar semantic structures. 

ChangeLog.md
    version 0.1.0 to find problem with building 
    and move to workspace11

    0.1.5.1  ghc 9.2.5 changed versino to be in sync with other packages

test/TestingTime.hs
-----------------------------------------------------------------------------
--
-- Module      :   top tests for time
-----------------------------------------------------------------------------
{-# OPTIONS_GHC -F -pgmF htfpp #-}
{-# OPTIONS_GHC -Wno-missing-signatures #-}

    {-# LANGUAGE
    MultiParamTypeClasses
    , TypeSynonymInstances
--    , FunctionalDependencies
    , FlexibleInstances
    , FlexibleContexts
    , ScopedTypeVariables
    , UndecidableInstances
    , OverloadedStrings
    , TypeFamilies

    #-}

module Main     where



import Data.Time 
import  {-@ HTF_TESTS @-} Uniform.Time_test
import Test.Framework
--import System.Time (calendarTimeToString)
-- import Text.Show.Pretty
import Uniform.Strings
import Uniform.Time as UT

main = do
    putIOwords ["HTF Time:\n "]
    r <- htfMainWithArgs [] -- ["--quiet"] 
                -- htf_thisModulesTests
                htf_importedTests
    putIOwords ["HTF end LayoutTest.hs:\n posTest"]
    return r

-- timeTest ::   IO [Bool]
-- timeTest = do
--     r <- runErr timeTest2
--     v1 <- case r of
--         Left msg -> do
--                 putIOwords ["errorTest returned Left :", msg]
--                 return [False]
--         Right v -> return v
--     return v1




-- timeTest2 ::   ErrIO [Bool]
-- -- examples how to use
-- timeTest2 = do
--     t1 <- getCurrentTimeUTC
--     putIOwords ["now",   showT t1]
--     let     t2 = addSeconds 4.0 t1
--             d4 = diffSeconds t2 t1
--             x1 = d4 == 4.0
--     putIOwords ["diff d4", showT d4, showT t2]

--     let day1 = toYMD t1


--     let     tomorrow = addSeconds (24 * 60 * 60) t1
--             day2 = toYMD tomorrow
--             oneDay = UT.diffDays  tomorrow t1
--             x2 = oneDay == 1

--     putIOwords ["today is", showT day1, "tomorrow", showT day2]


--     return [x1, x2, True]




test/Uniform/Time_test.hs
-----------------------------------------------------------------------------
--
-- Module      :   top tests for time
-----------------------------------------------------------------------------
{-# OPTIONS_GHC -F -pgmF htfpp #-}
{-# OPTIONS_GHC -Wno-missing-signatures #-}
    {-# LANGUAGE
    MultiParamTypeClasses
    , TypeSynonymInstances
   , FunctionalDependencies
    , FlexibleInstances
    , FlexibleContexts
    , ScopedTypeVariables
    , UndecidableInstances
    , OverloadedStrings
    , TypeFamilies

    #-}

module Uniform.Time_test     where



import Data.Time 
import Uniform.Time {-@ HTF_TESTS @-} 
import Test.Framework
--import System.Time (calendarTimeToString)
-- import Text.Show.Pretty
import Uniform.Strings
import Uniform.Time as UT hiding (ppShow)
-- import Uniform.Pretty 




test_date1 =
  do
    let dateString = "26 Jan 2012 10:54 AM"
    let timeFromString = parseTimeOrError True defaultTimeLocale "%d %b %Y %l:%M %p" dateString :: UTCTime
    -- Format YYYY/MM/DD HH:MM
    print $ formatTime defaultTimeLocale "%Y/%m/%d %H:%M" timeFromString
    -- Format MM/DD/YYYY hh:MM AM/PM
    let res = formatTime defaultTimeLocale "%m/%d/%Y %I:%M %p" timeFromString
    assertEqual "01/26/2012 10:54 AM" res

test_date2 = do
    -- now for a string with single digit months and days:
    let dateString = "9-8-2012 10:54 AM"
    let timeFromString = parseTimeOrError True defaultTimeLocale "%-d-%-m-%Y %l:%M %p" dateString :: UTCTime
    -- Format YYYY/MM/DD HH:MM
    let res = formatTime defaultTimeLocale "%Y/%m/%d %H:%M" timeFromString
--    let res = formatTime defaultTimeLocale "%b %d, %Y" timeFromString
    assertEqual "2012/08/09 10:54" res

test_date3 =
    assertEqual "2019-01-07 00:00:00 UTC" (showT $ readDate2 "Jan 7, 2019")

test_date4 = do
    let nowT = readDate2 "Jan 7, 2019"
    now <- getCurrentTime
    let res = formatTime defaultTimeLocale "%b %-d, %Y" nowT
    assertEqual "Jan 7, 2019" res

--test_date5= do
--    r1 :: ErrOrVal Text <- runErr $ getDateAsText
--    let r2 = fromRightNote "werw" r1
--    assertEqual "Mon Mar 11 17:37:12 CET 2019" r2

test_date3a =
    assertEqual r3a (map (showT . readDate3) t3a)

r3a =
    ["2019-01-03 00:00:00 UTC", "2019-01-03 00:00:00 UTC",
     "2019-03-05 00:00:00 UTC", "2019-01-06 00:00:00 UTC",
     "2011-02-07 00:00:00 UTC", "2018-01-01 00:00:00 UTC",
     "2020-10-09 00:00:00 UTC", "2018-11-27 00:00:00 UTC",
     "2009-05-14 00:00:00 UTC", "2019-01-04 00:00:00 UTC"]
t3a = ["Jan 3, 2019", "Jan. 3, 2019", "March 5, 2019"
    , "jan 6, 2019", "February 7, 2011"
    , "1.1.2018", "9.10.20", "2018-11-27", "2009-05-14"
    , "Jan. 4, 2019"]


test_showRead =
--    do
--        print time1
--        print show1
--        print read1
        assertEqual read1 time1

    where
     dateString = "9-8-2012 10:54 AM"
     time1 :: UTCTime
     time1 = parseTimeOrError True defaultTimeLocale "%-d-%-m-%Y %l:%M %p" dateString :: UTCTime
     show1 = show time1 :: String
     read1 = read show1 :: UTCTime

-- test_show = assertEqual (timeX) 
--         (readNoteT "test_showT" timeXshow :: UTCTime)
-- timeXshow = show timeX :: String 

-- test_pp = assertEqual (show timeX) (ppShow timeX)
-- demonstrates the issue with ppShow which cannot be 
-- read 

-- test_pp2 = assertEqual "" (ppShow timeX)
-- just to produce a display of "2012-08-09 10:54:00 UTC"

timeX :: UTCTime
timeX = parseTimeOrError True defaultTimeLocale "%-d-%-m-%Y %l:%M %p"
                "9-8-2012 10:54 AM" :: UTCTime

-- test_force = assertBool False 


Uniform/Time.hs
----------------------------------------------------------------------
--
-- Module      :  Uniform.Time
--
----------------------------------------------------------------------
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings #-}

-- | a minimal set of time operations
-- at the moment only a wrapper to time
-- examples in TestingTime.hs
module Uniform.Time
  ( module Uniform.Time,
    -- module Uniform.Strings,  
    EpochTime,
    UTCTime (..),
    ErrIO 
  )
where

import Data.Convertible (convert)
import Data.Time as T
  ( NominalDiffTime,
    UTCTime (..),
    addUTCTime,
    defaultTimeLocale,
    diffDays,
    diffUTCTime,
    formatTime,
    getCurrentTime,
    parseTimeM,
    parseTimeOrError,
    toGregorian,
  )
import Data.Time.Clock.POSIX
    ( getCurrentTime, posixSecondsToUTCTime )
import System.Posix.Types (EpochTime)

import Uniform.Error
--(ErrIO, errorT)

import Uniform.Strings
-- Text, CharChains2, show', IsString(..), t2s, s2)

import Uniform.Zero

year2000 :: UTCTime
year2000 = readDate3 "2000-01-01"
-- ^ may serve as zero in some applications

instance CharChains2 UTCTime Text where -- orphan instance
  show' = s2t . show

instance CharChains2 T.NominalDiffTime Text where
  show' = s2t . show

instance CharChains2 (Integer, Int, Int) Text where
  show' = s2t . show

instance IsString UTCTime where
  fromString = readNote "IsString UTCTime"

getCurrentTimeUTC :: ErrIO UTCTime
addSeconds :: Double -> UTCTime -> UTCTime
diffSeconds :: UTCTime -> UTCTime -> T.NominalDiffTime
getCurrentTimeUTC = liftIO T.getCurrentTime

addSeconds s t = T.addUTCTime (realToFrac s) t

diffSeconds = T.diffUTCTime

toYMD :: UTCTime -> (Integer, Int, Int)
toYMD = T.toGregorian . T.utctDay

diffDays :: UTCTime -> UTCTime -> Integer
diffDays a b = T.diffDays (T.utctDay a) (T.utctDay b)

epochTime2UTCTime :: EpochTime -> UTCTime
epochTime2UTCTime = convert

getDateAsText :: ErrIO Text
getDateAsText = callIO $ do
  now <- getCurrentTime
  let res = formatTime defaultTimeLocale "%b %-d, %Y" now
  return . s2t $ res


readDate2 :: Text -> UTCTime
-- ^ read data in the Jan 7, 2019 format (no . after month)
readDate2 datestring =
  parseTimeOrError
    True
    defaultTimeLocale
    "%b %-d, %Y"
    (t2s datestring) ::
    UTCTime

readDate3 :: Text -> UTCTime
readDate3 dateText = case (readDateMaybe dateText) of
  Nothing -> errorT ["readDate3", dateText, "cannot be parsed"]
  Just t -> t

readDateMaybe :: Text -> Maybe UTCTime
-- ^ read data in various formats (but not 9.10.20 !)
readDateMaybe dateText =
  listToMaybe . catMaybes $
    [ shortMonth,
      longMonth,
      monthPoint,
      germanNumeralShort,
      germanNumeral,
      isoformat
    ]
  where
    shortMonth :: Maybe UTCTime
    shortMonth =
      parseTimeM
        True
        defaultTimeLocale
        "%b %-d, %Y"
        dateString ::
        Maybe UTCTime
    longMonth =
      parseTimeM
        True
        defaultTimeLocale
        "%B %-d, %Y"
        dateString ::
        Maybe UTCTime
    monthPoint =
      parseTimeM
        True
        defaultTimeLocale
        "%b. %-d, %Y"
        dateString ::
        Maybe UTCTime
    germanNumeral =
      parseTimeM
        True
        defaultTimeLocale
        "%-d.%-m.%Y"
        dateString ::
        Maybe UTCTime
    germanNumeralShort =
      parseTimeM
        True
        defaultTimeLocale
        "%-d.%-m.%y"
        dateString ::
        Maybe UTCTime
    isoformat =
      parseTimeM
        True
        defaultTimeLocale
        "%Y-%m-%d"
        dateString ::
        Maybe UTCTime

    dateString = t2s dateText

fromEpochTime' :: EpochTime -> UTCTime
fromEpochTime' et = posixSecondsToUTCTime (realToFrac et)


src/TestingTime.hs
-----------------------------------------------------------------------------
--
-- Module      :   top tests for time
-----------------------------------------------------------------------------
{-# OPTIONS_GHC -F -pgmF htfpp #-}

    {-# LANGUAGE
    MultiParamTypeClasses
    , TypeSynonymInstances
--    , FunctionalDependencies
    , FlexibleInstances
    , FlexibleContexts
    , ScopedTypeVariables
    , UndecidableInstances
    , OverloadedStrings
    , TypeFamilies

    #-}

module Main     where


import Uniform.Strings

import {-@ HTF_TESTS @-} Uniform.Time

import Test.Framework

main = do
    putIOwords ["HTF LayoutTest.hs:\n posTest"]
    r <- htfMainWithArgs ["--quiet"] htf_importedTests
    putIOwords ["HTF end LayoutTest.hs:\n posTest"]
    return r

timeTest ::   IO [Bool]
timeTest = do
    r <- runErr timeTest2
    v1 <- case r of
        Left msg -> do
                putIOwords ["errorTest returned Left :", msg]
                return [False]
        Right v -> return v
    return v1




timeTest2 ::   ErrIO [Bool]
-- examples how to use
timeTest2 = do
    t1 <- getCurrentTimeUTC
    putIOwords ["now",   showT t1]
    let     t2 = addSeconds 4.0 t1
            d4 = diffSeconds t2 t1
            x1 = d4 == 4.0
    putIOwords ["diff d4", showT d4, showT t2]

    let day1 = toYMD t1


    let     tomorrow = addSeconds (24 * 60 * 60) t1
            day2 = toYMD tomorrow
            oneDay = diffDays  tomorrow t1
            x2 = oneDay == 1

    putIOwords ["today is", showT day1, "tomorrow", showT day2]


    return [x1, x2, True]




